docs/consistent-hash.md

```diff
@@ -0,0 +1,93 @@
+
+- The data of existing actors will not be displaced too much. 
+- The block cache of a compute node will not be invalidated too much.
+
+### Batch
+
+When we perform parallel batch read, we should partition the data for each parallelism in some way. Now that we have vnodes corresponding to disjoint sets of data, this naturally forms a data partition pattern: one vnode could be viewed as a minimal data partition unit, and we could aggregate several vnodes together to get a larger data partition.
+
+In vnode mapping introduced above, one parallel unit will correspond to several vnodes, so we could view vnodes that are mapped to one parallel unit as one partition group. In [the figure above](#data-distribution), namely, a partition group is vnodes with the same color (or the data that are hashed to these vnodes).
+
+This is better than range partition in that this approach of partition is more stable when the primary key of a materialized view distributes non-randomly, for example, monotonically increasing.
+
+### Storage
+
+If we look into the read-write pattern of streaming actors, we'll find that in most cases, actors only need to read the data written by itself (i.e. actor's internal states). Namely, read data with the same vnodes as it previously writes. 
+
+Therefore, an instinctive way to place data in storage is to **group data by vnodes**. In this way, when actors perform read operation, they could touch as few SST blocks as possible and thus trigger less I/O. 
+
+We know that [Hummock](./state-store-overview.md#overview), our LSM-Tree-based storage engine, sorts key-value pairs by the order of the key. Hence, in order to group data by vnode on the basis of Hummock, we **encode vnode into the storage key**. The storage key will look like
+```
+table_id | vnode | ...
+```
+where `table_id` denotes the [state table](./storing-state-using-relational-table.md#relational-table-layer), and `vnode` is computed via $H$ on key of the data. 
+
+To illustrate this, let's revisit the [previous example](#streaming). Executors of an operator will share the same logical state table, just as is shown in the figure below:
+
+![actor state table](./images/consistent-hash/actor-state-table.svg)
+
+Now that we have 12 vnodes in total in the example, the data layout in storage will accordingly look like this:
+
+![storage data layout](./images/consistent-hash/storage-data-layout.svg)
+
+Note that we only show the logical sequence and aggregation of data in this illustration. The actual data may be separated into different SSTs in Hummock.
+
+Since the way that certain data are hashed to vnode is invariant, the encoding of the data will also be invariant. How we schedule the fragment (e.g. parallelism of the fragment) will not affect data encoding. In other words, storage will not care about vnode mapping, which is determined by meta and used only by streaming. This is actually a way of decoupling compute layer and storage layer.
```

Comment:
Another minor rewording suggestion: the last sentence-----> "This is actually a way of decoupling the compute layer from the storage layer."